charger/vehicle_controlled_charger.go

```diff
@@ -0,0 +1,188 @@
+package charger
+
+import (
+	"errors"
+	"math"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/core/loadpoint"
+	"github.com/evcc-io/evcc/util"
+)
+
+// VehicleControlledCharger is a charger implementation that delegates control to the vehicle
+// This is useful for "granny chargers" or simple chargers that can't be controlled directly
+type VehicleControlledCharger struct {
+	log              *util.Logger
+	lp               loadpoint.API
+	enabled          bool
+	wasDisconnected  bool
+	geofenceEnabled  bool
+	chargerLatitude  float64
+	chargerLongitude float64
+	radius           float64
+}
+
+func init() {
+	registry.Add("vehicle-controlled", NewVehicleControlledChargerFromConfig)
+}
+
+// NewVehicleControlledChargerFromConfig creates a new vehicle-controlled charger
+func NewVehicleControlledChargerFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		GeofenceEnabled bool    `mapstructure:"geofence_enabled"`
+		Latitude        float64 `mapstructure:"latitude"`
+		Longitude       float64 `mapstructure:"longitude"`
+		Radius          float64 `mapstructure:"radius"`
+	}{
+		GeofenceEnabled: false,
+		Radius:          100, // Default 100 meter radius
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	c := &VehicleControlledCharger{
+		log:              util.NewLogger("vehicle-controlled"),
+		enabled:          false,
+		wasDisconnected:  true,
+		geofenceEnabled:  cc.GeofenceEnabled,
+		chargerLatitude:  cc.Latitude,
+		chargerLongitude: cc.Longitude,
+		radius:           cc.Radius,
+	}
+
+	return c, nil
+}
+
+// isVehicleAtCharger checks if the vehicle is within the geofence (if enabled)
+func (c *VehicleControlledCharger) isVehicleAtCharger(vehicle api.Vehicle) (bool, error) {
+	if !c.geofenceEnabled {
+		return true, nil // Assume at charger if geofencing is disabled
```

Comment:
You could check if the vehicle supports the status interface and use that here as fallback.

Reply:
Hmm, not sure if I get what you mean. I am already relying on the Status() interface, but that does not tell me if the vehicle is connected to THIS charging location.

Reply:
That's true. But if it's not connected to _any_ charger it certainly isn't connected to this one.

Reply:
If geofencing is not enabled, then vehicleIsAtCharger is always true, and the logic falls back to using the vehicle status interface only: 

https://github.com/evcc-io/evcc/blob/47f93b40227ae37e64af0b3169293fc9d771932b/charger/vehicle_controlled_charger.go#L104

So that's already using the status interface as fallback if geofence is disabled, is that what you meant?